rotkehlchen/tests/exchanges/test_binance.py

```diff
@@ -790,3 +770,70 @@ def mock_get_deposit_withdrawal(url):  # pylint: disable=unused-argument
                 end_ts=Timestamp(end_ts),
             )
             assert mock_api_query_dict.call_args_list == expected_calls
+
+
+@pytest.mark.freeze_time(datetime(2020, 11, 24, 3, 14, 15))
+def test_api_query_retry_on_status_code_429(function_scope_binance):
+    """Test when Binance API returns 429 and the request is retried, the
+    signature is not polluted by any attribute from the previous call.
+
+    It also tests getting the `retry-after` seconds to backoff from the
+    response header.
+
+    NB: basically remove `call_options['signature']`.
+    """
+    binance = function_scope_binance
+    offset_ms = 1000
+    call_options = {
+        'fromId': 0,
+        'limit': 1000,
+        'symbol': 'BUSDUSDT',
+        'recvWindow': 10000,
+        'timestamp': str(ts_now_in_ms() + offset_ms),
+    }
+    signature = hmac.new(
+        binance.secret,
+        urlencode(call_options).encode('utf-8'),
+        hashlib.sha256,
+    ).hexdigest()
+    call_options['signature'] = signature
+    base_url = 'https://api.binance.com/api/v3/myTrades?'
+    exp_request_url = base_url + urlencode(call_options)
+
+    # NB: all calls must have the same signature (time frozen)
+    expected_calls = [call(exp_request_url), call(exp_request_url), call(exp_request_url)]
+
+    def get_mocked_response():
+        responses = [
+            MockResponse(429, '[]', headers={'retry-after': '1'}),
+            MockResponse(418, '[]', headers={'retry-after': '5'}),
+            MockResponse(418, '[]', headers={'retry-after': str(RETRY_AFTER_LIMIT + 1)}),
+        ]
+        for response in responses:
+            yield response
+
+    def mock_response(url):  # pylint: disable=unused-argument
+        return next(get_response)
+
+    get_response = get_mocked_response()
+
+    # force to break the loop after 2 requests by lowering the `self.backoff_limit`
+    backoff_patch = patch.object(binance, 'backoff_limit', new=7)
```

Comment:
So @vnavascues here is something. If you study the patch documentation you will see that making 3 magic mocks and then putting them as callables in a patch is unnecessarily complicated.

You patch what you want and then you can decide what to patch. The first 2 you change their value so you use the kwarg `new` and for the last one you change the function so you use the kwarg `side_effect`. No need to create 3 magic mocks.

See the diff frrom before here: https://github.com/rotki/rotki/pull/1860/commits/52ff983af04b04160fd75df930f007c38ddde010